### PR TITLE
fix: controller.topologySpreadConstraints schema

### PR DESCRIPTION
**Pull request by marcuz**
### Proposed changes
Fixes error:

```
controller.topologySpreadConstraints: Invalid type. Expected: object, given: array
```

Expected YAML:

```yaml
controller:
  topologySpreadConstraints:
    - maxSkew: 1
      topologyKey: topology.kubernetes.io/zone
      whenUnsatisfiable: DoNotSchedule
      labelSelector:
        matchLabels:
          app: nginx-ingress
    - maxSkew: 1
      topologyKey: kubernetes.io/hostname
      whenUnsatisfiable: DoNotSchedule
      labelSelector:
        matchLabels:
          app: nginx-ingress
```

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/main/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto main
- [x] I will ensure my PR is targeting the main branch and pulling from my branch from my own fork

### CODE REVIEW

**Review by vepatel — APPROVED** [review]

**Review by jjngx — APPROVED** [review]

File: deployments/helm-chart/values.schema.json [hunk 1]
```diff
@@ -455,8 +455,8 @@
           "$ref": "https://raw.githubusercontent.com/yannh/kubernetes-json-schema/master/v1.25.4/_definitions.json#/definitions/io.k8s.api.core.v1.Affinity"
         },
         "topologySpreadConstraints": {
-          "type": "object",
-          "default": {},
+          "type": "array",
+          "default": [],
           "title": "The topologySpreadConstraints Schema",
           "$ref": "https://raw.githubusercontent.com/yannh/kubernetes-json-schema/master/v1.25.4/_definitions.json#/definitions/io.k8s.api.core.v1.TopologySpreadConstraint"
```
Comment (lucacome) [inline deployments/helm-chart/values.schema.json:461]:
```suggestion
          "$ref": "https://raw.githubusercontent.com/yannh/kubernetes-json-schema/master/v1.25.4/_definitions.json#/definitions/io.k8s.api.core.v1.PodSpec/properties/topologySpreadConstraints"
```

**Review by lucacome — CHANGES_REQUESTED** [review]
Hi @marcuz I think we just need to change the `ref` to the right definition and everything else can remain the same

File: deployments/helm-chart/values.schema.json (hunk 1 above)
Comment (marcuz) [reply deployments/helm-chart/values.schema.json:461]: Apologies, I missed that!

File: deployments/helm-chart/README.md [hunk 2]
```diff
@@ -177,7 +177,7 @@ Parameter | Description | Default
 `controller.terminationGracePeriodSeconds` | The termination grace period of the Ingress Controller pod. | 30
 `controller.tolerations` | The tolerations of the Ingress Controller pods. | []
 `controller.affinity` | The affinity of the Ingress Controller pods. | {}
-`controller.topologySpreadConstraints` | The topology spread constraints of the Ingress controller pods. | {}
+`controller.topologySpreadConstraints` | The topology spread constraints of the Ingress controller pods. | []
```
Comment (lucacome) [inline deployments/helm-chart/README.md:180]:
```suggestion
`controller.topologySpreadConstraints` | The topology spread constraints of the Ingress controller pods. | {}
```

File: deployments/helm-chart/values.schema.json [hunk 3]
```diff
@@ -1283,7 +1283,7 @@
           },
           "tolerations": [],
           "affinity": {},
-          "topologySpreadConstraints": {},
+          "topologySpreadConstraints": [],
```
Comment (lucacome) [inline deployments/helm-chart/values.schema.json:1286]:
```suggestion
          "topologySpreadConstraints": {},
```

File: deployments/helm-chart/values.schema.json [hunk 4]
```diff
@@ -1584,7 +1584,7 @@
         },
         "tolerations": [],
         "affinity": {},
-        "topologySpreadConstraints": {},
+        "topologySpreadConstraints": [],
```
Comment (lucacome) [inline deployments/helm-chart/values.schema.json:1587]:
```suggestion
        "topologySpreadConstraints": {},
```

File: deployments/helm-chart/values.yaml [hunk 5]
```diff
@@ -151,7 +151,7 @@ controller:
   affinity: {}
 
   ## The topology spread constraints of the Ingress controller pods.
-  # topologySpreadConstraints: {}
+  # topologySpreadConstraints: []
```
Comment (lucacome) [inline deployments/helm-chart/values.yaml:154]:
```suggestion
  # topologySpreadConstraints: {}
```

**Comment by marcuz** [comment]
@lucacome sorry I totally misunderstood the behaviour of `$ref` here. Pushing a cleanup commit soon!

**Comment by marcuz** [comment]
Just changing the `$ref` as suggested, lint passes successfully:

```
$ helm lint . -f values.yaml -f values-test.yaml 
==> Linting .

1 chart(s) linted, 0 chart(s) failed
```

`values-test.yaml` contains the `controller.topologySpreadConstraints` as specified above.

**Review by lucacome — APPROVED** [review]
Thanks @marcuz !

**Comment by marcuz** [comment]
> Thanks @marcuz !

Sorry for all the back and forth. :smile:

**Comment by cpuengr949** [comment]
@lucacome I am getting this same error using terraform to build
`      topologySpreadConstraints = [
        {
          labelSelector = {
            matchLabels = {
              app = "nginx-plus-nginx-ingress"
            }
          }
          maxSkew           = 1
          topologyKey       = "kubernetes.io/hostname"
          whenUnsatisfiable = "DoNotSchedule"
         }
       ]
`
Which is creating the code
`                  "topologySpreadConstraints":
                  - "labelSelector":
                      "matchLabels":
                        "app": "nginx-plus-nginx-ingress"
                    "maxSkew": 1
                    "topologyKey": "kubernetes.io/hostname"
                    "whenUnsatisfiable": "DoNotSchedule"
`

**Comment by lucacome** [comment]
@cpuengr949 even with the fix in this PR?

**Comment by cpuengr949** [comment]
@lucacome How do I get the fix? 
I believe terraform downloads it each time right?
Thank You

**Comment by lucacome** [comment]
@cpuengr949 if it's pulling the `main` branch from here or using the `edge` helm chart, it should have the fix in this PR, but I don't know what your setup is 🙂

**Pull request merged**